src/main/java/net/minecraftforge/server/command/TagsCommand.java

```diff
@@ -0,0 +1,266 @@
+        /*
+         * /forge tags <registry> list [page]
+         * /forge tags <registry> get <tag> [page]
+         * /forge tags <registry> query <element> [page]
+         */
+        return Commands.literal("tags")
+                .then(Commands.argument("registry", ResourceKeyArgument.key(ROOT_REGISTRY_KEY))
+                        .suggests(TagsCommand::suggestRegistries)
+                        .then(Commands.literal("list")
+                                .executes(ctx -> listTags(ctx, 1))
+                                .then(Commands.argument("page", IntegerArgumentType.integer(1))
+                                        .executes(ctx -> listTags(ctx, IntegerArgumentType.getInteger(ctx, "page")))
+                                )
+                        )
+                        .then(Commands.literal("get")
+                                .then(Commands.argument("tag", ResourceLocationArgument.id())
+                                        .suggests(suggestFromRegistry(r -> r.getTagNames().map(TagKey::location)::iterator))
+                                        .executes(ctx -> listTagElements(ctx, 1))
+                                        .then(Commands.argument("page", IntegerArgumentType.integer(1))
+                                                .executes(ctx -> listTagElements(ctx, IntegerArgumentType.getInteger(ctx, "page")))
+                                        )
+                                )
+                        )
+                        .then(Commands.literal("query")
+                                .then(Commands.argument("element", ResourceLocationArgument.id())
+                                        .suggests(suggestFromRegistry(Registry::keySet))
+                                        .executes(ctx -> queryElementTags(ctx, 1))
+                                        .then(Commands.argument("page", IntegerArgumentType.integer(1))
+                                                .executes(ctx -> queryElementTags(ctx, IntegerArgumentType.getInteger(ctx, "page")))
+                                        )
+                                )
+                        )
+                );
+    }
+
+    private static int listTags(final CommandContext<CommandSourceStack> ctx, final int page) throws CommandSyntaxException
+    {
+        final ResourceKey<? extends Registry<?>> registryKey = getResourceKey(ctx, "registry", ROOT_REGISTRY_KEY)
+                .orElseThrow(); // Expect to be always retrieve a resource key for the root registry (registry key)
+        final Registry<?> registry = ctx.getSource().getServer().registryAccess().registry(registryKey)
+                .orElseThrow(() -> UNKNOWN_REGISTRY.create(registryKey.location()));
+
+        final long tagCount = registry.getTags().count();
+
+        ctx.getSource().sendSuccess(createMessage(
+                new TranslatableComponent("%s", new TextComponent(registryKey.location().toString()).withStyle(ChatFormatting.GOLD)),
```

Comment:
What is with the random TranslatableTextComponent of just `%s`? Why doesn't this just pass the created `TextComponent` directly and then change `createMessage` to take a `MutableComponent` instead of a `TranslatableComponent`

Reply:
@sciwhiz12

Reply:
This was a remnant of when I refactored the common code between `listTags` and `listTagElements` into the separate helper method `createMessage`. As can be seen in the other methods, `TranslatableComponent` is used because of `%s / %s` for listing tag elements and `%s : %s` for listing tags which hold a certain element.

Additionally, the use of `TranslatableComponent` allows flexibility in changing the format of the command's output at either the programmer's discretion (by changing the string literal) or the final translator's discretion (through the translation keys), which would not otherwise be attainable if passing the text component directly.

Reply:
A couple counter points to your mention of flexibility:
1. The programmer can always wrap it in a translatable component if they want to change it further in the future, that isn't a great reason to add an extra wrapper around it in the meantime.
2. This is a more valid reason, except that the key isn't exposed as a lang entry? So there is no way for the final translator to adjust it; at which point having it doesn't gain anything.